src/com/facebook/buck/jvm/groovy/GroovyTestDescription.java

```diff
@@ -0,0 +1,193 @@
+import com.facebook.buck.jvm.java.TestType;
+import com.facebook.buck.model.BuildTarget;
+import com.facebook.buck.parser.NoSuchBuildTargetException;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildRuleResolver;
+import com.facebook.buck.rules.BuildRuleType;
+import com.facebook.buck.rules.BuildRules;
+import com.facebook.buck.rules.BuildTargetSourcePath;
+import com.facebook.buck.rules.Description;
+import com.facebook.buck.rules.Hint;
+import com.facebook.buck.rules.Label;
+import com.facebook.buck.rules.SourcePath;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.rules.TargetGraph;
+import com.facebook.buck.util.HumanReadableException;
+import com.facebook.infer.annotation.SuppressFieldNotInitialized;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
+
+import java.nio.file.Path;
+import java.util.logging.Level;
+
+public class GroovyTestDescription implements Description<GroovyTestDescription.Arg> {
+
+  public static final BuildRuleType TYPE = BuildRuleType.of("groovy_test");
+
+  private final GroovyBuckConfig groovyBuckConfig;
+  private final JavacOptions defaultJavacOptions;
+  private final Optional<Long> defaultTestRuleTimeoutMs;
+  private final Optional<Path> testTempDirOverride;
+
+  public GroovyTestDescription(
+      GroovyBuckConfig groovyBuckConfig,
+      JavacOptions defaultOptions,
+      Optional<Long> defaultTestRuleTimeoutMs,
+      Optional<Path> testTempDirOverride) {
+    this.groovyBuckConfig = groovyBuckConfig;
+    this.defaultJavacOptions = defaultOptions;
+    this.defaultTestRuleTimeoutMs = defaultTestRuleTimeoutMs;
+    this.testTempDirOverride = testTempDirOverride;
+  }
+
+  @Override
+  public BuildRuleType getBuildRuleType() {
+    return TYPE;
+  }
+
+  @Override
+  public Arg createUnpopulatedConstructorArg() {
+    return new Arg();
+  }
+
+  @Override
+  public <A extends Arg> JavaTest createBuildRule(
+      TargetGraph targetGraph,
+      BuildRuleParams params,
+      BuildRuleResolver resolver,
+      A args) throws NoSuchBuildTargetException {
+    SourcePathResolver pathResolver = new SourcePathResolver(resolver);
+
+    BuildTarget abiJarTarget =
+        BuildTarget.builder(params.getBuildTarget())
+            .addFlavors(CalculateAbi.FLAVOR)
+            .build();
+
+    GroovycToJarStepFactory stepFactory = new GroovycToJarStepFactory(
+        groovyBuckConfig.getGroovyCompiler().get(),
+        args.extraGroovycArguments,
+        JavacOptionsFactory.create(
+            defaultJavacOptions,
+            params,
+            resolver,
+            pathResolver,
+            args
+        ));
+
+    JavaTest test =
+        resolver.addToIndex(
```

Comment:
You don't need to add `test` to the resolver since you're returning it from `createBuildRule`, it will get added by https://github.com/facebook/buck/blob/master/src/com/facebook/buck/rules/TargetGraphToActionGraph.java#L99.